IGraph.h

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2004-2015 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+//===----------------------------------------------------------------------===//
+// Chapel LLVM Locality Optimization by Akihiro Hayashi (ahayashi@rice.edu)
+//===----------------------------------------------------------------------===//
+// SSA Value Graph for Locality Inference
+//===----------------------------------------------------------------------===//
+
+#ifndef _IGRAPH_H_
+#define _IGRAPH_H_
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/DOTGraphTraits.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvmUtil.h"
+
+#if HAVE_LLVM_VER >= 35
+#else
+#include "llvm/Assembly/Writer.h"
+#endif
+
+#include <vector>
+#include <string>
+using namespace std;
+using namespace llvm;
+
+class Node {
+private:
+    StringRef name;
+    Value* value;
+    vector<Node*> children;
+    vector<Node*> parents;
+    DenseMap<Value*, unsigned> LLMap;
+    unsigned int getAddressSpace(Value *v);
+    void initLLMap();
+
+public:
+    Node(Value* _value) { value = _value; initLLMap(); };
+
+    // parents
+    vector<Node*>::iterator parents_begin() { return parents.begin(); }
+    vector<Node*>::iterator parents_end() { return parents.end(); }
+    // children
+    vector<Node*>::iterator begin() { return children.begin(); }
+    vector<Node*>::iterator end() { return children.end(); }
+    vector<Node*>::const_iterator begin() const { return children.begin(); }
+    vector<Node*>::const_iterator end() const { return children.end(); }
+
+    int getLL(Value* v) const {
```

Comment:
Could you add comments about what this is doing? What does LL indicate? What is the purpose of the LL map? what is this function returning?